src/user-support.c

```diff
@@ -123,11 +53,13 @@ void setup_user_xdg_runtime_dir()
 	}
 
 	errno = 0;
-	mkpath(xdg_runtime_dir);
-
-	// if successfully created the directory (ie, not EEXIST), then
-	// chmod it.
+	debug("creating user XDG_RUNTIME_DIR directory: %s", xdg_runtime_dir);
+	if (sc_nonfatal_mkpath(xdg_runtime_dir, 0755) < 0) {
+		die("cannot create user XDG_RUNTIME_DIR directory: %s",
+		    xdg_runtime_dir);
+	}
+	// if successfully created the directory (ie, not EEXIST), then chmod it.
 	if (errno == 0 && chmod(xdg_runtime_dir, 0700) != 0) {
-		die("could not chmod XDG_RUNTIME_DIR");
+		die("cannot change permissions of user XDG_RUNTIME_DIR directory to 0700");
```

Comment:
Note there is a bug here that was present before this PR. /run/user/UID is supposed to be 0700. XDG_RUNTIME_DIR is set to /run/user/UID/snap.name. If /run/user/UID didn't exist and snap-confine created it, then it will be 0755. This could be solved by something along the lines of (note `man 1 dirname` for use of tmp):

```
// XDG_RUNTIME_DIR as set by 'snap run' is a subdirectory of the user's session
// XDG_RUNTIME_DIR (ie, /run/user/UID/snap.name). Since it is possible that the
// user's session XDG_RUNTIME_DIR (ie, /run/user/UID) does not exist yet, follow
// the freedesktop.org specification and ensure the session /run/user/UID is 0700
// when we create it ourselves, and also make sure the snap's XDG_RUNTIME_DIR
// is 0700.
char *tmp __attribute__ ((cleanup(sc_cleanup_string))) = NULL;
char *parent_dir __attribute__ ((cleanup(sc_cleanup_string))) = NULL;
￼tmp = strdup(xdg_runtime_dir);
parent_dir = dirname(tmp);
if (sc_nonfatal_mkpath(parent_dir, 0755) < 0) {
...
if (errno == 0 && chmod(parent_dir, 0700) != 0) {
...
if (sc_nonfatal_mkpath(xdg_runtime_dir, 0755) < 0) {
...
if (errno == 0 && chmod(xdg_runtime_dir, 0700) != 0) {
...
```